discovery/endorsement/endorsement.go

```diff
@@ -150,23 +171,51 @@ func (ea *endorsementAnalyzer) computeEndorsementResponse(ctx *context) (*discov
 
 	layouts := computeLayouts(ctx.principalsSets, principalGroups, satGraph)
 	if len(layouts) == 0 {
-		return nil, errors.New("cannot satisfy any principal combination")
+		return nil, errors.New("no peer combination can satisfy the endorsement policy")
 	}
 
 	criteria := &peerMembershipCriteria{
-		possibleLayouts: layouts,
-		satGraph:        satGraph,
-		chanMemberById:  ctx.channelMembersById,
-		idOfMembers:     ctx.identitiesOfMembers,
+		possibleLayouts:  layouts,
+		satGraph:         satGraph,
+		chanMemberById:   ctx.channelMembersById,
+		idOfMembers:      ctx.identitiesOfMembers,
+		chaincodeMapping: ctx.chaincodeMapping,
+	}
+
+	groupToEndorserListMapping := endorsersByGroup(criteria)
+	layouts = filterOutUnsatisfiedLayouts(groupToEndorserListMapping, layouts)
+
+	if len(layouts) == 0 {
+		return nil, errors.New("required chaincodes are not installed on sufficient peers")
 	}
 
 	return &discovery.EndorsementDescriptor{
 		Chaincode:         ctx.chaincode,
 		Layouts:           layouts,
-		EndorsersByGroups: endorsersByGroup(criteria),
+		EndorsersByGroups: groupToEndorserListMapping,
 	}, nil
 }
 
+func filterOutUnsatisfiedLayouts(endorsersByGroup map[string]*discovery.Peers, layouts []*discovery.Layout) []*discovery.Layout {
+	// Iterate once again over all layouts and ensure every layout has enough peers in the EndorsersByGroups
+	// as required by the quantity in the layout.
+	filteredLayouts := make([]*discovery.Layout, len(layouts))
```

Comment:
don't you have to create it as
`filteredLayouts := make([]*discovery.Layout, 0, len(layouts))`
? Otherwise you're just appending to a bunch of nils on line 214

Reply:
Do you see now why i prefer my slices, cut thin?

Reply:
done, thanks for noticing.